simplify SourceID Hash

Eh2406 opened the pull request:
### What does this PR try to resolve?

Despite being interned `SourceId::Eq` is not a `ptr::eq`. Which in turn is because `SourceId`s concept of identity is a complete mess. The code relies on having to IDs that are `Eq` but do not have the same values for their fields.

As one measure of this `SourceId` has an `impl Hash` which does something different from `fn full_hash` and `fn stable_hash`. Separately `SourceIdInner` has a different implementation. Similar levels of complexity exist for `Eq`. Every one of these `impl`s  was added due to a real bug/issue we've had that needs to stay fixed. Not all of witch are reproducible enough to have made it into our test suite.

I [have some ideas](https://github.com/rust-lang/cargo/pull/14665#issuecomment-2412077472) for how to reorganize the types so that this is easier to reason about and faster. But given the history and the complexity I want to move extremely carefully.

### How should we test and review this PR?

The test pass, and it's a one line change,  but this still needs careful review.

### Additional information

r? @ehuss I remember you and Alex working very hard to track down most of these bugs.

Eh2406 commented:
If this turns out not to be a valid transformation I am happy to update the PR to be tests to prove why it's not okay.

epage commented on src/cargo/core/source_id.rs:
```diff
@@ -666,15 +666,12 @@ impl fmt::Display for SourceId {
-            _ => self.inner.url.as_str().hash(into),
-        }
+        self.inner.canonical_url.hash(into);
     }
```
Most CanonicalUrl differences are related to git somehow.  I'm assuming that means we can't hit them in any other way.  There is the stripping of a trailing `/`.  However, we probably should have been respecting that in the hash in the first place I assume.  The question is if this will change anything significant enough

Eh2406 replied:
Perhaps if you have a index hosted on git, then `GITHUB.com/me/my-index.git/` could change a lot. I think the biggest concern here is having to recheck out dependencies. So even if this does affect people we might just eat that cost. On the other hand there's a lot of complexity here and maybe there are bigger problems I forgot about.

weihanglo replied:
Haven't yet looked closer. I wonder if this would affect hacks like this one: <https://github.com/rust-lang/cargo/issues/5478#issuecomment-1055365283>.

Eh2406 replied:
I don't think that would be affected by this change. But it is something we have to watch for as I "clean up" this code.

weihanglo replied:
That is true, since `[patch.<url>]` uses URLs directly, not SourceId.

This change may affect anywhere we let people define URL and parse it as `SourceId`, like the `[registries]` table? We could also deem this as a bug and fix it.

Eh2406 replied:
I would hope that we rely on `Eq` for those cases and not `Hash`. So that the problem you identified is not a problem with this PR, rather one to investigate on the next PR that changes `Eq`.

weihanglo replied:
Fair enough. Have checked other usages of it, none of them seems user-facing like the path to index files. It should be fine to move over.

weihanglo reviewed (approved):
Since the paths to crates.io Git and spare index on users disks are not affected by this change, you have my vote.

ehuss commented:
To be clear, the goal here is to simplify the code a little bit? Are there any other intended improvements?

And the downside is that if someone has a custom registry that canonicalizes differently (like a trailing slash, or upper-case characters in a GitHub URL), it will hash differently and they will end up with new directories in `~/.cargo/registry/*/...`, and duplicated cache contents?

If that's the tradeoff, I'm a little on the fence and could go either way.

If I'm understanding it correctly, would it be possible to update the `test_stable_hash` test with an example that illustrates the hash changing with this PR?

Eh2406 commented:
>>    And the downside is that if someone has a custom registry that canonicalizes differently (like a trailing slash, or upper-case characters in a GitHub URL), it will hash differently and they will end up with new directories in ~/.cargo/registry/*/..., and duplicated cache contents?
> 
> That is the only concern I'm aware of. But it is possible that I missed something.

While trying to update the tests as requested I thought of another concern.  If the user has two custom registries that canonical eyes to the same thing, before this PR they would end up in separate directories in `~/.cargo/registry/*/...` and now they would end up in the same folder. With this collision cause problems?

weihanglo commented:
> , before this PR they would end up in separate directories in `~/.cargo/registry/*/...` and now they would end up in the same folder. With this collision cause problems?

Theoretically yes. However I doubt somebody relies on this behavior to distinguish custom registries. It either needs to have trailing `/` or `.git`. or on gitub.com with protocol scheme other than `https`. We could notify interest groups in advance. Not sure what is the correct channel, [t-cargo/build-integration](https://rust-lang.zulipchat.com/#narrow/channel/334885-t-cargo.2Fbuild-integration) doesn't seem to be the best fit.

bors commented:
:umbrella: The latest upstream changes (presumably a8f7db2af913b634ab52fc554c7f7c36804c7de0) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Eh2406 commented:
I was going to abandon this PR, but before I do... @weihanglo do those filenames  still go through this code path at all? Where do they go through `stable_hash`?

weihanglo commented:
> I was going to abandon this PR, but before I do... @weihanglo do those filenames still go through this code path at all? Where do they go through `stable_hash`?

The code path doesn't change at all. `stable_hash` still relies on Hash impl for SourceId

https://github.com/rust-lang/cargo/blob/652623b779c88fe44afede28bf7f1c9c07812511/src/cargo/core/source_id.rs?plain=1#L553

Eh2406 commented:
Good point! But we can just in line that.  I just rebased and added a commit.  The first commit changes `stable_hash` to not depend on `hash`  but to have its own copy of the code. Then the second commit simplifies the main `hash`.

I feel like this alleviates the relevant concerns.

weihanglo commented:
> Good point! But we can just in line that. […] I feel like this alleviates the relevant concerns.

That is true. However I am forgetting the motivation of this PR. It doesn't seem to simplify the current code. Does this change open any door for future improvement?

Eh2406 commented:
> It doesn't seem to simplify the current code. 

Separating the "must return the same hash value across versions of the code" into 1  function, even though there's more code, I do think it is a very small when for clarity. The Hash trait does not normally have that requirement, our reliance on it always feels a little awkward.

> I am forgetting the motivation of this PR. [..] Does this change open any door for future improvement?

Yes, and I'm happy to post a draft PR of where this is going if that would help. The idea is that  I like to separately intern the parts of SourceID that are part of `Eq` from the parts that are mutable. In order for that to be effective, I need to minimize the parts that are relevant to `Eq`. But, `Hash == Hash  must imply Eq == Eq`  so in order to simplify `Eq`  I needed to start by simplifying `Hash`. I think the three changes deserve their own PRs is because:
- this PR, the change to Hash, is mostly worried about the stable hash interactions.
- the next PR, the change to Eq,  It is mostly concerned with the interactions between the interning strategy and its implications on  the rest of the code.
- the last PR, changes to interning, are mostly concerned about performance.
If I did it in one PR, I'd be very concerned that  one or more of these concerns might get lost in the other conversations.

Opinions can and do vary. I am open to however you wish to proceed.

weihanglo reviewed (approved):
Thanks for the explanation, Jacob!
